Adding deprecated to @Available directive

mustiikhalil opened the pull request:
## Summary

Adding `deprecated` argument to the directive `@Available`, this is done by adding a field `deprecated` to the `Metadata.Availability` and also addressing all the comments that are related to the issue within the codebase.

This PR also include the `.any` platform for `@Available`, which can be used as follows:

```
@Available(*, introduced: "1.0")
@Available(*, introduced: "1.0", deprecated: "10.0")
```

Closes #441 

Testing

Adding the following metadata into `SwiftDocC/Directives.md`

```
@Metadata {
    @Available("Swift DocC", introduced: "1.0", deprecated: "2.0")
}
```

Using [swift-docc-render-artifact](https://github.com/apple/swift-docc-render-artifact)

Steps:

1. Add the above metadata to `Sources/SwiftDocC/SwiftDocC.docc/SwiftDocC/Directives.md`.
2. `./bin/preview-docs SwiftDocC`

3. Ensure that in the documentation list the "Directives" shows a deprecated tag
<img width="414" alt="Screenshot 2024-03-08 at 4 09 05 PM" src="https://github.com/apple/swift-docc/assets/26250654/36730c30-c9b2-4fec-9398-4d48041f7c0c">

4. Ensure that the "Directives" page is printed with an availability callout for "Swift DocC" which contains a introduced and deprecated versions

<img width="738" alt="Screenshot 2024-03-08 at 4 09 12 PM" src="https://github.com/apple/swift-docc/assets/26250654/9734cc1e-ba78-437f-b165-605770696561">

## Checklist

- [X] Added tests
- [X] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary 
   (Not sure how to update the docs properly)

sofiaromorales commented:
@mustiikhalil Could you please add the Steps to reproduce section to the PR template, including the expected behavior and a test DocC catalog that uses the new deprecated keyword?

mustiikhalil commented:
> Could you please add the Steps to reproduce section to the PR template, including the expected behaviour

Yes, i can do that later today.

@sofiaromorales regarding the test bundle i thought it would be enough to include it within the current test bundle which can be found [here](Tests/SwiftDocCTests/Test Bundles/AvailabilityBundle.docc/AvailableArticle.md) since its just an arguement to the same directive

mustiikhalil commented:
@sofiaromorales sorry for the ping, but would love to know what's next to get this PR moving forward

sofiaromorales commented:
> @sofiaromorales sorry for the ping, but would love to know what's next to get this PR moving forward

Sorry for the wait. I left some initial feedback and in the upcoming days I'll do another round of reviews.

sofiaromorales commented:
> Updated documentation if necessary
(Not sure how to update the docs properly)

@mustiikhalil please update the `Availability` DocC comment to properly document this change.

mustiikhalil commented:
I'll look at it during the weekend.

mustiikhalil commented on Sources/SwiftDocC/Semantics/Metadata/Availability.swift:
```diff
@@ -77,6 +74,7 @@ extension Metadata {
                 case .iOS: return "iOS"
                 case .watchOS: return "watchOS"
                 case .tvOS: return "tvOS"
+                case .any: return "*"
```
It would look something like the screenshot below:

<img width="743" alt="Screenshot 2024-03-22 at 5 33 26 PM" src="https://github.com/apple/swift-docc/assets/26250654/c8aa1081-9a3b-4f25-81e1-e6322324470a">

d-ronnqvist replied:
I don't know how this is meant to work but I don't think we'd want this to display `*` here. 

It's a bit odd that the wildcard specifies a version at all because that version may not align across the various platforms.

mustiikhalil replied:
If that's the case what would be the correct representation to say that it is deprecated on all platforms?

Where as a specific case of deprecation might look like this for example:

<img width="744" alt="Screenshot 2024-03-28 at 10 10 36 PM" src="https://github.com/apple/swift-docc/assets/26250654/cc7000ad-657b-4aed-9a33-902a86de6b24">

d-ronnqvist replied:
I would probably say that the correct way would be to list the deprecation on platform where the symbol is known to be available. 

However, like I said above, it's a bit odd that we allow specifying a deprecation version for all platforms since the different platforms don't need to align their versions and in practice don't do so.

mustiikhalil replied:
Ah that makes sense, I'll update the PR as soon as possible with removing the `*`. And thus aligning with screenshot sent below/ in the previous comment:

<img width="744" alt="Screenshot 2024-03-28 at 10 10 36 PM" src="https://github.com/apple/swift-docc/assets/26250654/cc7000ad-657b-4aed-9a33-902a86de6b24">

mustiikhalil replied:
Should be fixed now

mustiikhalil commented on Tests/SwiftDocCTests/Test Bundles/AvailabilityBundle.docc/ComplexAvailableDeprecated.md:
```diff
@@ -0,0 +1,13 @@
+# Complex Available Deprecated Article
```
`ComplexAvailableDeprecated.md` I guess can be replaced with the already existing tests in `testValidDirective`  since it looks like this:

```swift
        let source = """
         @Metadata {
             @Available(macOS, introduced: "11.0")
             @Available(iOS, introduced: "15.0")
             @Available(watchOS, introduced: "7.0", deprecated: "9.0")
             @Available(*, introduced: "0.1", deprecated: "1.0")
         }
         """
         try assertValidMetadata(source: source)
 ```

mustiikhalil commented:
@d-ronnqvist what can I do to make this PR move forward?

d-ronnqvist commented:
> @d-ronnqvist what can I do to make this PR move forward?

Sorry for the very, very long delay. 

In the Documentation Workgroup meeting on June 3, one of the topics that we discussed was how "*" availability should be displayed on the page to better reflect that a large number of packages work on Linux and Windows and anywhere else where Swift is supported. We didn't come to any specific conclusions about how it should be displayed but decided that it should be pitched on the Swift Forums where the broader community can be part of the conversation. 

The goal is to discuss and hopefully implement those changes in the 6.0 timeframe but it may still take a few weeks before everyone agrees on how it should be displayed and how it should behave. 

If you don't want to wait for that discussion to happen, it may be good to separate the `any` platform changes and limit this PR to the new support for the "deprecated" argument to the `@Available` directive.

mustiikhalil commented:
> Sorry for the very, very long delay.

No worries

> If you don't want to wait for that discussion to happen, it may be good to separate the any platform changes and limit this PR to the new support for the "deprecated" argument to the @Available directive.

If that's the case let's keep this PR tiny, as in adding support for deprecated. And create an issue for the "any" which I would be happy to work on when the Specs for it are ready. 

I will simply revert the changes when parsing the "*" operator in the platforms enum. And revert a couple of the tests, comments that removed the mention of `.any/*`

What do you think?

mustiikhalil commented on Tests/SwiftDocCTests/Semantics/MetadataAvailabilityTests.swift:
```diff
@@ -90,17 +87,12 @@ class MetadataAvailabilityTests: XCTestCase {
             try assertValidAvailability(source: "@Available(\"My Package\", \(args))")
         }
 
-        // also test for giving no platform
-        for args in validArguments {
-            try assertValidAvailability(source: "@Available(\(args))")
-        }
```
Removed this code since it was never called because `validArguments` was always empty before. However now we have `deprecated` as a value within the array

mustiikhalil commented:
> That sounds like a good plan

@d-ronnqvist Super sorry for the late update, but this should be good now. Only added a simple parameter and reverted all the instances where I removed the wild card "*" comments.

d-ronnqvist commented:
@swift-ci please test
